File: python-zeladoria/app/routes/comentarios.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List
from pydantic import BaseModel
from datetime import datetime

from app.database.database import get_db
from app.models.comentario import Comentario
from app.models.chamado import Chamado
from app.models.usuario import Usuario
from app.utils.auth import get_current_user

router = APIRouter(prefix="/api/chamados/{chamado_id}/comentarios")

# ============= SCHEMAS =============

class ComentarioCreate(BaseModel):
    comentario: str
    tipo: str = "atualizacao"  # atualizacao, observacao, resolucao
    visivel_cidadao: bool = True

class ComentarioResponse(BaseModel):
    id: int
    chamado_id: int
    usuario_id: int
    usuario_nome: str
    usuario_tipo: str
    comentario: str
    tipo: str
    visivel_cidadao: bool
    data_criacao: str

    class Config:
        from_attributes = True

# ============= ENDPOINTS =============

@router.post("", response_model=dict)
def adicionar_comentario(
    chamado_id: int,
    comentario_data: ComentarioCreate,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_current_user)
):
    """
    Adicionar comentário a um chamado
    
    Tipos de comentário:
    - atualizacao: Atualização sobre o andamento
    - observacao: Observação/nota interna ou externa
    - resolucao: Comentário sobre a resolução
    
    Permissões:
    - Equipe, Gestor, Secretaria, Admin: podem comentar
    - Cidadão: pode comentar no próprio chamado
    """
    
    # Verificar se chamado existe
    chamado = db.query(Chamado).filter(Chamado.id == chamado_id).first()
    if not chamado:
        raise HTTPException(status_code=404, detail="Chamado não encontrado")
    
    # Verificar permissões
    if current_user.tipo == "cidadao" and chamado.usuario_id != current_user.id:
        raise HTTPException(
            status_code=403, 
            detail="Você só pode comentar em seus próprios chamados"
        )
    
    # Criar comentário
    novo_comentario = Comentario(
        chamado_id=chamado_id,
        usuario_id=current_user.id,
        comentario=comentario_data.comentario,
        tipo=comentario_data.tipo,
        visivel_cidadao=1 if comentario_data.visivel_cidadao else 0
    )
    
    db.add(novo_comentario)
    db.commit()
    db.refresh(novo_comentario)
    
    return {
        "success": True,
        "message": "Comentário adicionado com sucesso",
        "comentario": novo_comentario.to_dict()
    }

@router.get("", response_model=List[ComentarioResponse])
def listar_comentarios(
    chamado_id: int,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_current_user)
):
    """
    Listar todos os comentários de um chamado
    
    Regras de visibilidade:
    - Cidadão: vê apenas comentários com visivel_cidadao=True
    - Equipe/Gestor/Secretaria/Admin: vê todos os comentários
    """
    
    # Verificar se chamado existe
    chamado = db.query(Chamado).filter(Chamado.id == chamado_id).first()
    if not chamado:
        raise HTTPException(status_code=404, detail="Chamado não encontrado")
    
    # Buscar comentários
    query = db.query(Comentario).filter(Comentario.chamado_id == chamado_id)
    
    # Filtrar por visibilidade se for cidadão
    if current_user.tipo == "cidadao":
        query = query.filter(Comentario.visivel_cidadao == 1)
    
    comentarios = query.order_by(Comentario.data_criacao.asc()).all()
    
    # Converter para resposta
    return [
        ComentarioResponse(
            id=c.id,
            chamado_id=c.chamado_id,
            usuario_id=c.usuario_id,
            usuario_nome=c.usuario.nome if c.usuario else "Sistema",
            usuario_tipo=c.usuario.tipo if c.usuario else "sistema",
            comentario=c.comentario,
            tipo=c.tipo,
            visivel_cidadao=bool(c.visivel_cidadao),
            data_criacao=c.data_criacao.isoformat() if c.data_criacao else None
        )
        for c in comentarios
    ]

@router.delete("/{comentario_id}")
def deletar_comentario(
    chamado_id: int,
    comentario_id: int,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_current_user)
):
    """
    Deletar um comentário
    
    Permissões:
    - Admin: pode deletar qualquer comentário
    - Gestor: pode deletar comentários do próprio setor
    - Usuário: pode deletar apenas seus próprios comentários (até 30 min)
    """
    
    comentario = db.query(Comentario).filter(
        Comentario.id == comentario_id,
        Comentario.chamado_id == chamado_id
    ).first()
    
    if not comentario:
        raise HTTPException(status_code=404, detail="Comentário não encontrado")
    
    # Verificar permissões
    pode_deletar = False
    
    if current_user.tipo == "admin":
        pode_deletar = True
    elif current_user.tipo == "gestor":
        pode_deletar = True  # Gestor pode deletar qualquer comentário
    elif comentario.usuario_id == current_user.id:
        # Usuário pode deletar seu próprio comentário apenas se for recente
        tempo_decorrido = (datetime.utcnow() - comentario.data_criacao).total_seconds()
        if tempo_decorrido <= 1800:  # 30 minutos
            pode_deletar = True
        else:
            raise HTTPException(
                status_code=403,
                detail="Você só pode deletar comentários até 30 minutos após a criação"
            )
    
    if not pode_deletar:
        raise HTTPException(
            status_code=403,
            detail="Você não tem permissão para deletar este comentário"
        )
    
    db.delete(comentario)
    db.commit()
    
    return {"success": True, "message": "Comentário deletado com sucesso"}

@router.patch("/{comentario_id}")
def editar_comentario(
    chamado_id: int,
    comentario_id: int,
    comentario_data: ComentarioCreate,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_current_user)
):
    """
    Editar um comentário existente
    
    Permissões:
    - Usuário pode editar seu próprio comentário (até 30 min)
    - Admin/Gestor podem editar qualquer comentário
    """
    
    comentario = db.query(Comentario).filter(
        Comentario.id == comentario_id,
        Comentario.chamado_id == chamado_id
    ).first()
    
    if not comentario:
        raise HTTPException(status_code=404, detail="Comentário não encontrado")
    
    # Verificar permissões
    pode_editar = False
    
    if current_user.tipo in ["admin", "gestor"]:
        pode_editar = True
    elif comentario.usuario_id == current_user.id:
        # Usuário pode editar seu próprio comentário apenas se for recente
        tempo_decorrido = (datetime.utcnow() - comentario.data_criacao).total_seconds()
        if tempo_decorrido <= 1800:  # 30 minutos
            pode_editar = True
        else:
            raise HTTPException(
                status_code=403,
                detail="Você só pode editar comentários até 30 minutos após a criação"
            )
    
    if not pode_editar:
        raise HTTPException(
            status_code=403,
            detail="Você não tem permissão para editar este comentário"
        )
    
    # Atualizar comentário
    comentario.comentario = comentario_data.comentario
    comentario.tipo = comentario_data.tipo
    comentario.visivel_cidadao = 1 if comentario_data.visivel_cidadao else 0
    
    db.commit()
    db.refresh(comentario)
    
    return {
        "success": True,
        "message": "Comentário atualizado com sucesso",
        "comentario": comentario.to_dict()
    }

File: python-zeladoria/frontend/comentarios.js
// ========================================
// 💬 SISTEMA DE COMENTÁRIOS
// ========================================

async function carregarComentarios(chamadoId) {
    try {
        const response = await apiRequest(`/chamados/${chamadoId}/comentarios`);
        
        if (!response.ok) {
            console.error('Erro ao carregar comentários');
            return [];
        }
        
        const comentarios = await response.json();
        return comentarios;
    } catch (error) {
        console.error('Erro ao carregar comentários:', error);
        return [];
    }
}

function renderizarComentarios(comentarios, chamadoId) {
    if (comentarios.length === 0) {
        return `
            <div style="text-align: center; padding: 30px; color: var(--secondary);">
                <div style="font-size: 3rem; margin-bottom: 10px;">💬</div>
                <p>Nenhum comentário ainda</p>
                <p style="font-size: 0.9rem;">Seja o primeiro a comentar sobre este chamado!</p>
            </div>
        `;
    }
    
    return comentarios.map(c => {
        const data = new Date(c.data_criacao);
        const dataFormatada = formatDate(data);
        
        // Ícones por tipo de usuário
        const iconeUsuario = {
            'cidadao': '👤',
            'equipe': '👷',
            'gestor': '📊',
            'secretaria': '🏛️',
            'admin': '🔧',
            'sistema': '🤖'
        }[c.usuario_tipo] || '👤';
        
        // Cor por tipo de comentário
        const corTipo = {
            'atualizacao': '#dbeafe',
            'observacao': '#fef3c7', 
            'resolucao': '#d1fae5'
        }[c.tipo] || '#f1f5f9';
        
        const labelTipo = {
            'atualizacao': '📝 Atualização',
            'observacao': '💡 Observação',
            'resolucao': '✅ Resolução'
        }[c.tipo] || '💬 Comentário';
        
        const podeEditar = currentUser && 
                          (c.usuario_id === currentUser.id || 
                           ['admin', 'gestor'].includes(currentUser.tipo));
        
        return `
            <div class="comentario-item" style="background: white; border-left: 4px solid ${corTipo}; padding: 15px; border-radius: 8px; margin-bottom: 15px;">
                <div style="display: flex; justify-content: space-between; align-items: start; margin-bottom: 10px;">
                    <div>
                        <div style="display: flex; align-items: center; gap: 8px; margin-bottom: 5px;">
                            <span style="font-size: 1.2rem;">${iconeUsuario}</span>
                            <strong>${c.usuario_nome}</strong>
                            <span style="background: ${corTipo}; padding: 2px 8px; border-radius: 4px; font-size: 0.75rem; font-weight: 600;">
                                ${labelTipo}
                            </span>
                            ${!c.visivel_cidadao ? '<span style="background: #fee2e2; color: #991b1b; padding: 2px 8px; border-radius: 4px; font-size: 0.75rem; font-weight: 600;">🔒 Interno</span>' : ''}
                        </div>
                        <div style="font-size: 0.85rem; color: var(--secondary);">
                            ${dataFormatada}
                        </div>
                    </div>
                    ${podeEditar ? `
                    <div style="display: flex; gap: 5px;">
                        <button onclick="editarComentario(${chamadoId}, ${c.id}, '${c.comentario.replace(/'/g, "\\'")}', '${c.tipo}', ${c.visivel_cidadao})" 
                                class="btn-icon" title="Editar">
                            ✏️
                        </button>
                        <button onclick="deletarComentario(${chamadoId}, ${c.id})" 
                                class="btn-icon" title="Deletar">
                            🗑️
                        </button>
                    </div>
                    ` : ''}
                </div>
                <div style="white-space: pre-wrap; color: var(--dark); line-height: 1.5;">
                    ${c.comentario}
                </div>
            </div>
        `;
    }).join('');
}

async function adicionarComentario(chamadoId) {
    const comentarioTexto = document.getElementById('novo-comentario-texto').value.trim();
    const tipo = document.getElementById('novo-comentario-tipo').value;
    const visivelCidadao = document.getElementById('novo-comentario-visivel').checked;
    
    if (!comentarioTexto) {
        showToast('Digite um comentário', 'warning');
        return;
    }
    
    showLoading();
    
    try {
        const response = await apiRequest(`/chamados/${chamadoId}/comentarios`, {
            method: 'POST',
            body: JSON.stringify({
                comentario: comentarioTexto,
                tipo: tipo,
                visivel_cidadao: visivelCidadao
            })
        });
        
        if (response.ok) {
            showToast('Comentário adicionado com sucesso!');
            
            // Limpar formulário
            document.getElementById('novo-comentario-texto').value = '';
            document.getElementById('novo-comentario-tipo').value = 'atualizacao';
            document.getElementById('novo-comentario-visivel').checked = true;
            
            // Recarregar comentários
            await atualizarComentariosModal(chamadoId);
        } else {
            const data = await response.json();
            showToast(data.detail || 'Erro ao adicionar comentário', 'error');
        }
    } catch (error) {
        showToast('Erro: ' + error.message, 'error');
    } finally {
        hideLoading();
    }
}

async function atualizarComentariosModal(chamadoId) {
    const comentarios = await carregarComentarios(chamadoId);
    const container = document.getElementById('comentarios-container');
    
    if (container) {
        container.innerHTML = renderizarComentarios(comentarios, chamadoId);
    }
}

async function deletarComentario(chamadoId, comentarioId) {
    if (!confirm('Tem certeza que deseja deletar este comentário?')) {
        return;
    }
    
    showLoading();
    
    try {
        const response = await apiRequest(`/chamados/${chamadoId}/comentarios/${comentarioId}`, {
            method: 'DELETE'
        });
        
        if (response.ok) {
            showToast('Comentário deletado com sucesso!');
            await atualizarComentariosModal(chamadoId);
        } else {
            const data = await response.json();
            showToast(data.detail || 'Erro ao deletar comentário', 'error');
        }
    } catch (error) {
        showToast('Erro: ' + error.message, 'error');
    } finally {
        hideLoading();
    }
}

async function editarComentario(chamadoId, comentarioId, comentarioAtual, tipoAtual, visivelAtual) {
    const novoComentario = prompt('Editar comentário:', comentarioAtual);
    
    if (novoComentario === null || novoComentario.trim() === '') {
        return;
    }
    
    showLoading();
    
    try {
        const response = await apiRequest(`/chamados/${chamadoId}/comentarios/${comentarioId}`, {
            method: 'PATCH',
            body: JSON.stringify({
                comentario: novoComentario.trim(),
                tipo: tipoAtual,
                visivel_cidadao: visivelAtual
            })
        });
        
        if (response.ok) {
            showToast('Comentário atualizado com sucesso!');
            await atualizarComentariosModal(chamadoId);
        } else {
            const data = await response.json();
            showToast(data.detail || 'Erro ao atualizar comentário', 'error');
        }
    } catch (error) {
        showToast('Erro: ' + error.message, 'error');
    } finally {
        hideLoading();
    }
}

// Adicionar CSS para botões de ícone
const style = document.createElement('style');
style.textContent = `
    .btn-icon {
        background: none;
        border: none;
        cursor: pointer;
        font-size: 1rem;
        padding: 4px 8px;
        border-radius: 4px;
        transition: all 0.2s;
    }
    
    .btn-icon:hover {
        background: var(--light);
        transform: scale(1.1);
    }
    
    .comentarios-secao {
        border-top: 2px solid var(--border);
        margin-top: 20px;
        padding-top: 20px;
    }
    
    .form-comentario {
        background: var(--light);
        padding: 15px;
        border-radius: 8px;
        margin-bottom: 20px;
    }
`;
document.head.appendChild(style);
